In [1]:
from Neuron import *
from NeuronLayer import *
from NeuronNetwork import *

from sklearn.model_selection import train_test_split
from sklearn import datasets
import pandas as pd
import random
import time

## Prep the data

In [2]:
digits = datasets.load_digits()
digits.target = pd.get_dummies(digits.target)

In [3]:
inputs = []
targets = []
for i in range(len(digits.data)):
    inputs.append(list(digits.data[i]))
    targets.append(list(digits.target.iloc[i].values))

## Normaliseren van de input data

In [4]:
inputs = normalize2DList(inputs)

### Split train en test set

In [5]:
inputsTrain, inputsTest, targetsTrain, targetsTest = train_test_split(
inputs, targets, test_size=0.33, random_state=42)

## Create neuron layers

In [6]:
hiddenNeurons = []
for i in range(8):
    weights = [random.randrange(-3, 3) for xyz in range(64)]
    hiddenNeurons.append(Neuron(bias=0, weights=weights))
hiddenLayer = NeuronLayer(neurons=hiddenNeurons)

In [7]:
hiddenNeurons2 = []
for i in range(4):
    weights = [random.randrange(-3, 3) for xyz in range(8)]
    hiddenNeurons2.append(Neuron(bias=0, weights=weights))
hiddenLayer2 = NeuronLayer(neurons=hiddenNeurons2)

In [8]:
outputNeurons = []
for i in range(10):  # Output layer heeft 10 output neuronen
    weights = [random.randrange(-3, 3) for xyz in range(4)]
    outputNeurons.append(Neuron(bias=0, weights=weights))
outputLayer = NeuronLayer(neurons=outputNeurons)

## Create network

In [9]:
network = NeuronNetwork([hiddenLayer, hiddenLayer2, outputLayer])

## Train the network

In [10]:
startTime = time.time()
network.fit(inputs=inputsTrain, targets=targetsTrain, epochs=150, learningRate=0.4)
print("seconds: ", (time.time() - startTime))

seconds:  159.89257550239563


## Results

In [11]:
print("Test score: ", network.score(inputsTest, targetsTest), "%")
print("Train score: ", network.score(inputsTrain, targetsTrain), "%")
print("MSE: ", network.MSE)

Test score:  90.74074074074075 %
Train score:  96.17622610141314 %
MSE:  0.07697617520246931


Een mooie score behaald, als ik een hogere score wilt moet ik gaan werken met een kleinere learning rate en meer epochs. Om dit elke keer te proberen kost het 160 seconden. Dit duurt mij te lang, het gaat erom dat die kan trainen en dat het werkt!

## Using batches

In [11]:
network = NeuronNetwork([hiddenLayer, hiddenLayer2, outputLayer])

In [12]:
startTime = time.time()
network.fit(inputs=inputsTrain, targets=targetsTrain, epochs=150, learningRate=0.4, batchSize=16)
print("seconds: ", (time.time() - startTime))

seconds:  97.7369921207428


Met een batchSize van 16 dan traint die wat sneller, omdat je niet na elk voorbeeld alles update bespaart dit tijd.

## Results

In [13]:
print("Test score: ", network.score(inputsTest, targetsTest), "%")
print("Train score: ", network.score(inputsTrain, targetsTrain), "%")
print("MSE: ", network.MSE)

Test score:  89.73063973063972 %
Train score:  94.43059019118868 %
MSE:  0.1017552546345047


Redelijk de zelfde score als met een batchSize van 1